File: sort revision/sort_three_types.cpp
#include<iostream>
using namespace std;

void sortThree(int arr[],int n)
{
    int low = 0,mid = 0,high = n-1;

    while(mid<=high)
    {
        if(arr[mid]==0)
        {
            swap(arr[low],arr[mid]);
            low++;
            mid++;
        }
        else if(arr[mid]==1)
        {
            mid++;
        }
        else
        {
            swap(arr[mid],arr[high]);
            high--;
        }
    }
}

void display(int arr[],int n)
{
    for (int i = 0; i < n; i++)
    {
        cout<<arr[i]<<" ";
    }
    
}

int main()
{
    int arr[] = {0,2,1,2,2,0};
    int n = 6;

    sortThree(arr,n);
    display(arr,n);
    return 0;
}

File: sort revision/sort_three_types.js
let swap = (arr, low, high) => {
  let temp = arr[low];
  arr[low] = arr[high];
  arr[high] = temp;
};

let sort_three_types = (arr) => {
  let low = 0,
    mid = 0,
    high = arr.length - 1;

  while (mid < high) {
    if (arr[mid] === 0) {
      swap(arr, low, mid);
      low++;
      mid++;
    } else if (arr[mid] === 1) {
      mid++;
    } else {
      swap(arr, mid, high);
      high--;
    }
  }
  console.log(arr);
};

let arr = [0, 1, 2, 1, 1, 2];

sort_three_types(arr);
